Source/GSCSTableau.m

```diff
@@ -0,0 +1,700 @@
+                 overPivotableVariable: (GSCSVariable *)rhs
+{
+  return [lhs id] < [rhs id];
+}
+
+- (BOOL) hasColumnForVariable: (GSCSVariable *)variable
+{
+  return [_columns objectForKey: variable] != nil;
+}
+
+- (NSSet *) columnForVariable: (GSCSVariable *)variable
+{
+  return [_columns objectForKey: variable];
+}
+
+// Add v=expr to the tableau, update column cross indices
+// v becomes a basic variable
+- (void) addRowForVariable: (GSCSVariable *)variable
+          equalsExpression: (GSCSLinearExpression *)expression
+{
+  [_rows setObject: expression forKey: variable];
+  if ([variable isExternal])
+    {
+      [_externalRows setObject: expression forKey: variable];
+    }
+  [self addTermVariablesForExpression: expression variable: variable];
+}
+
+- (void) addTermVariablesForExpression: (GSCSLinearExpression *)expression
+                              variable: (GSCSVariable *)variable
+{
+  NSArray *termVariables = [expression termVariables];
+  FOR_IN(GSCSVariable *, expressionTermVariable, termVariables)
+    [self addMappingFromExpressionVariable: expressionTermVariable
+                             toRowVariable: variable];
+    if ([expressionTermVariable isExternal])
+      {
+        [_externalParametricVariables addObject: expressionTermVariable];
+      }
+  END_FOR_IN(termVariables);
+}
+
+- (void) addMappingFromExpressionVariable: (GSCSVariable *)columnVariable
+                            toRowVariable: (GSCSVariable *)rowVariable
+{
+  NSMutableSet *columnSet = [_columns objectForKey: columnVariable];
+  if (columnSet == nil)
+    {
+      columnSet = [NSMutableSet set];
+      [_columns setObject: columnSet forKey: columnVariable];
+    }
+  [columnSet addObject: rowVariable];
+}
+
+- (void) removeMappingFromExpressionVariable: (GSCSVariable *)columnVariable
+                               toRowVariable: (GSCSVariable *)rowVariable
+{
+  NSMutableSet *columnSet = [_columns objectForKey: columnVariable];
+  if (columnSet == nil)
+    {
+      return;
+    }
+  [columnSet removeObject: rowVariable];
+}
+
+- (void) removeColumn: (GSCSVariable *)variable
+{
+  NSSet *crows = [_columns objectForKey: variable];
+  if (_rows != nil)
+    {
+      FOR_IN(GSCSVariable *, clv, crows)
+        GSCSLinearExpression *expression = [_rows objectForKey: clv];
+        [expression removeVariable: variable];
+      END_FOR_IN(crows);
+      [_columns removeObjectForKey: variable];
+    }
+
+  if ([variable isExternal])
+    {
+      [_externalRows removeObjectForKey: variable];
+    }
+}
+
+- (void) removeRowForVariable: (GSCSVariable *)variable
+{
+  GSCSLinearExpression *expression = [_rows objectForKey: variable];
+  if (expression == nil)
+    {
+      NSException *missingExpressionException = [NSException
+        exceptionWithName: NSInvalidArgumentException
+                   reason: @"No expression exists for the provided variable"
+                 userInfo: nil];
+      [missingExpressionException raise];
+      return;
```

Comment:
Why a `return` here, the raise should never return.

Reply:
Fixed